cosmwasm/main.go

```diff
@@ -0,0 +1,280 @@
+	gob.Register(plugin.ExecutedHook{})
+}
+
+type p struct {
+	chainName string
+}
+
+// NewPlugin creates a new plugin instance with the given chainName and chainNamed values.
+func NewPlugin(chainName string) *p {
+	return &p{
+		chainName: chainName,
+	}
+}
+
+// ChainName returns the chainName of the plugin.
+func (p *p) ChainName() string {
+	return p.chainName
+}
+
+// ChainNamed derives and returns the chainNamed of the plugin.
+func (p *p) ChainNamed() string {
+	return p.chainName + "d"
+}
+func (p) Manifest() (plugin.Manifest, error) {
+	return plugin.Manifest{
+		Name: "cosmwasm",
+		// Add commands here
+		Commands: []plugin.Command{
+			// Example of a command
+			{
+				Use:               "cosmwasm",
+				Short:             "The cosmwasm command is used for adding cosmwasm support for apps scaffolded with Ignite CLI",
+				PlaceCommandUnder: "ignite",
+				// Examples of adding subcommands:
+				Commands: []plugin.Command{
+					{
+						Use:               "add",
+						Short:             "Add a CosmWasm module to the chain",
+						Long:              "This command will install dependencies, create new files, and modify existing files to add a CosmWasm module to the chain.",
+						PlaceCommandUnder: "ignite cosmwasm",
+					},
+				},
+			},
+		},
+	}, nil
+}
+
+func (p *p) Execute(cmd plugin.ExecutedCommand) error {
+
+	// According to the number of declared commands, you may need a switch:
+
+	switch cmd.Use {
+	case "add":
+		return handleAddCommand(p)
+	default:
+		return fmt.Errorf("unknown command: %s", cmd.Use)
+	}
+}
+
+func handleAddCommand(p *p) error {
+	fmt.Println("Adding stuff...")
+
+	err := installDependencies()
+	if err != nil {
+		return err
+	}
+
+	err = createNewFiles(p)
+	if err != nil {
+		return err
+	}
+
+	//err = modifyExistingFiles()
+	//if err != nil {return err}
+
+	return nil
+}
+
+func (p) ExecuteHookPre(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func (p) ExecuteHookPost(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func (p) ExecuteHookCleanUp(hook plugin.ExecutedHook) error {
+	return nil
+}
+
+func installDependencies() error {
+	cmd := exec.Command("go", "get", "github.com/CosmWasm/wasmd/x/wasm@v0.41.0")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("failed to install dependency: %w", err)
+	}
+	return nil
+}
+
+func replaceWordsInFile(filePath, targetWord, replacement string) error {
+	// Read content from the file
+	content, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return err
+	}
+
+	// Perform the replacement
+	modifiedContent := strings.ReplaceAll(string(content), targetWord, replacement)
+
+	// Write the modified content back to the file
+	err = ioutil.WriteFile(filePath, []byte(modifiedContent), 0644)
+	if err != nil {
+		return err
+	}
+
+	//fmt.Printf("Replaced planet with %s\n", replacement)
+
+	return nil
+}
+
+func createFile(inputFilename, outputDir, outputFilename string, p *p) error {
```

Comment:
I think here you can just send the `chainName` so you don't pass a complex type like the plugin. That way you keep the function API simpler and easy to unit test.

```suggestion
func createFile(inputFilename, outputDir, outputFilename, chainName string) error {
```

Reply:
resolved: https://github.com/yerasyla/plugins/commit/27f2900b6c853a13a1269fa3d74fe471399bc399